File: scripts/build_geo.mjs
/* Baja los límites oficiales y arma data/amba.geojson.
   Corre igual en local (`npm run build`) que en el build de Vercel.

   Fuentes: IGN (WFS de departamentos) y datos abiertos del GCBA (barrios).
   Los puntos de referencia, el cordón y la cobertura salen de scripts/zonas.json,
   donde ya están resueltos (ver build_geo.py para cómo se calcularon). */

import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import simplify from '@turf/simplify';

const AQUI = dirname(fileURLToPath(import.meta.url));
const RAIZ = join(AQUI, '..');

const IGN = 'https://wms.ign.gob.ar/geoserver/ows?service=WFS&version=1.0.0'
  + '&request=GetFeature&typeName=ign:departamento&outputFormat=application/json'
  + "&CQL_FILTER=in1%20LIKE%20'06%25'";

const GCBA = 'https://cdn.buenosaires.gob.ar/datosabiertos/datasets'
  + '/ministerio-de-educacion/barrios/barrios.geojson';

// 0,0001° ~ 11 m: el borde queda fiel al oficial incluso con mucho zoom.
const TOLERANCIA = 0.0001;

const norm = s => s.normalize('NFD').replace(/[̀-ͯ]/g, '').toLowerCase().trim();

async function bajar(url, que) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`${que}: HTTP ${res.status}`);
  return res.json();
}

function recortar(geom, nd = 5) {
  const r = o => Array.isArray(o) ? o.map(r) : Math.round(o * 10 ** nd) / 10 ** nd;
  return { ...geom, coordinates: r(geom.coordinates) };
}

const zonas = JSON.parse(await readFile(join(AQUI, 'zonas.json'), 'utf8'));

// Saavedra y San Nicolás son barrio porteño y partido bonaerense a la vez,
// así que cada fuente busca sólo entre las zonas de su región (cordón 0 = CABA).
const indice = { CABA: new Map(), GBA: new Map() };
for (const n of Object.keys(zonas)) {
  indice[zonas[n][2] === 0 ? 'CABA' : 'GBA'].set(norm(n), n);
}

const [partidos, barrios] = await Promise.all([
  bajar(IGN, 'IGN departamentos'),
  bajar(GCBA, 'GCBA barrios'),
]);

const salida = [];
const vistos = new Set();

function agregar(nombre, geometry, region, comunaFuente) {
  const [lat, lon, cordon, enRed, comuna] = zonas[nombre];
  const g = simplify({ type: 'Feature', properties: {}, geometry },
                     { tolerance: TOLERANCIA, highQuality: true, mutate: true });
  const props = { nombre, region, cordon, enRed: !!enRed, lat, lon };
  if (region === 'CABA') props.comuna = comuna ?? comunaFuente;
  salida.push({ type: 'Feature', properties: props, geometry: recortar(g.geometry) });
  vistos.add(nombre);
}

for (const f of partidos.features) {
  const nombre = indice.GBA.get(norm(f.properties.nam));
  if (nombre) agregar(nombre, f.geometry, 'GBA');
}

for (const f of barrios.features) {
  const nombre = indice.CABA.get(norm(f.properties.nombre));
  if (nombre) agregar(nombre, f.geometry, 'CABA', f.properties.comuna);
}

const faltan = Object.keys(zonas).filter(n => !vistos.has(n));
if (faltan.length) throw new Error('zonas sin geometría: ' + faltan.join(', '));

await mkdir(join(RAIZ, 'data'), { recursive: true });
const ruta = join(RAIZ, 'data', 'amba.geojson');
await writeFile(ruta, JSON.stringify({ type: 'FeatureCollection', features: salida }));

const enRed = salida.filter(f => f.properties.enRed).length;
console.log(`zonas: ${salida.length} | en la red: ${enRed} | ${(await readFile(ruta)).length} bytes`);

File: scripts/build_geo.py
# -*- coding: utf-8 -*-
"""Genera data/amba.geojson: los barrios de CABA y los partidos del GBA que
cubre la red de ProCourrier, con el punto de referencia de cada uno.

Fuentes: datos abiertos del GCBA (barrios) e IGN (departamentos).
"""
import json, os, unicodedata
from shapely.geometry import shape, mapping, Polygon
from shapely.ops import unary_union

SCRATCH = os.environ.get('PC_SCRATCH', '.')
OUT = os.path.join(os.path.dirname(__file__), '..', 'data')

# Partidos que la red alcanza, con su cordón. La red llega hasta Luján,
# Dique Luján (Tigre), Zárate, Campana y La Plata por el sur.
CORDON = {
    "Avellaneda": 1, "Lanús": 1, "Lomas de Zamora": 1, "La Matanza": 1, "Morón": 1,
    "Tres de Febrero": 1, "General San Martín": 1, "Vicente López": 1, "San Isidro": 1,
    "Hurlingham": 1, "Ituzaingó": 1,
    "Quilmes": 2, "Berazategui": 2, "Florencio Varela": 2, "Almirante Brown": 2,
    "Esteban Echeverría": 2, "Ezeiza": 2, "Merlo": 2, "Moreno": 2, "San Miguel": 2,
    "José C. Paz": 2, "Malvinas Argentinas": 2, "Tigre": 2, "San Fernando": 2,
    "Pilar": 3, "Escobar": 3, "General Rodríguez": 3, "Marcos Paz": 3, "Cañuelas": 3,
    "San Vicente": 3, "Presidente Perón": 3, "La Plata": 3, "Berisso": 3, "Ensenada": 3,
    "Campana": 3, "Zárate": 3, "Luján": 3,
}

# Partidos que se dibujan para dar contexto geográfico pero no tienen servicio.
FUERA_DE_RED = ["Exaltación de la Cruz", "Brandsen", "General Las Heras", "Navarro",
                "San Andrés de Giles"]

# El centroide geométrico de estos partidos cae en zona despoblada (el Delta,
# el sur rural de La Plata): el reparto se concentra en la cabecera.
CABECERA = {
    "San Fernando":      (-34.4472, -58.5702),
    "Tigre":             (-34.4235, -58.5818),
    "La Plata":          (-34.9207, -57.9538),
    "Campana":           (-34.1637, -58.9587),
    "San Vicente":       (-35.0249, -58.4241),
    "Marcos Paz":        (-34.7865, -58.8297),
    "Pilar":             (-34.4571, -58.9142),
}

# La red cobra distinto el norte y el sur de La Matanza. El corte es la
# mediatriz entre San Justo (norte) y el sudoeste del partido.
LM_NORTE = (-34.6774, -58.5608)
LM_SUR   = (-34.8327, -58.7011)

def norm(s):
    return "".join(c for c in unicodedata.normalize("NFD", s) if unicodedata.category(c) != "Mn").lower()

def redondear(o, nd=4):
    if isinstance(o, (list, tuple)): return [redondear(x, nd) for x in o]
    if isinstance(o, float): return round(o, nd)
    return o

def partir_la_matanza(geom):
    """Corta el partido en dos por la mediatriz entre el norte y el sur."""
    (ay, ax), (by, bx) = LM_NORTE, LM_SUR
    mx, my = (ax + bx) / 2, (ay + by) / 2
    dx, dy = bx - ax, by - ay
    largo = (dx * dx + dy * dy) ** .5
    nx, ny = dx / largo, dy / largo          # hacia el sur
    ux, uy = -ny, nx                         # sobre la línea de corte
    L = 5.0
    p1 = (mx + ux * L, my + uy * L)
    p2 = (mx - ux * L, my - uy * L)
    norte = Polygon([p1, p2, (p2[0] - nx * L, p2[1] - ny * L), (p1[0] - nx * L, p1[1] - ny * L)])
    sur   = Polygon([p1, p2, (p2[0] + nx * L, p2[1] + ny * L), (p1[0] + nx * L, p1[1] + ny * L)])
    g = shape(geom).buffer(0)
    return g.intersection(norte), g.intersection(sur)

def salida(nombre, g, region, cordon, en_red, ref=None, **extra):
    g = g.simplify(0.0015 if region == "GBA" else 0.0006, preserve_topology=True)
    p = ref or (lambda c: (c.y, c.x))(g.representative_point())
    props = {"nombre": nombre, "region": region, "cordon": cordon, "enRed": en_red,
             "lat": round(p[0], 5), "lon": round(p[1], 5), **extra}
    return {"type": "Feature", "properties": props, "geometry": redondear(mapping(g))}

def main():
    partidos = json.load(open(f"{SCRATCH}/ba_dptos.json"))
    barrios = json.load(open(f"{SCRATCH}/caba_barrios.geojson"))

    quiero = {norm(k): (k, v, True) for k, v in CORDON.items()}
    quiero.update({norm(k): (k, 3, False) for k in FUERA_DE_RED})
    feats, faltan = [], set(quiero)

    for f in partidos["features"]:
        k = norm(f["properties"]["nam"])
        if k not in quiero: continue
        faltan.discard(k)
        nombre, cordon, en_red = quiero[k]

        if nombre == "La Matanza":
            norte, sur = partir_la_matanza(f["geometry"])
            feats.append(salida("La Matanza Norte", norte, "GBA", 1, True, ref=LM_NORTE))
            feats.append(salida("La Matanza Sur", sur, "GBA", 2, True, ref=LM_SUR))
            continue

        feats.append(salida(nombre, shape(f["geometry"]).buffer(0), "GBA", cordon, en_red,
                            ref=CABECERA.get(nombre)))

    if faltan:
        print("NO ENCONTRADOS:", faltan)

    for f in barrios["features"]:
        p = f["properties"]
        feats.append(salida(p["nombre"].title(), shape(f["geometry"]).buffer(0),
                            "CABA", 0, True, comuna=p["comuna"]))

    os.makedirs(OUT, exist_ok=True)
    ruta = os.path.join(OUT, "amba.geojson")
    with open(ruta, "w", encoding="utf-8") as fh:
        json.dump({"type": "FeatureCollection", "features": feats}, fh,
                  ensure_ascii=False, separators=(",", ":"))

    en_red = sum(1 for f in feats if f["properties"]["enRed"])
    print(f"zonas: {len(feats)} | en la red: {en_red} | fuera: {len(feats) - en_red}")
    print(f"bytes: {os.path.getsize(ruta)}")

main()
